### PR TITLE
[Merged by Bors] - feat: HahnSeries on Lex product

### PR DESCRIPTION
**Pull request by ScottCarnahan**
This PR implements an equivalence between `HahnSeries Γ (HahnSeries Γ' R)` and `HahnSeries (Γ ×ₗ Γ') R`.  We do this by showing that a subset of the lexicographic product is partially well-ordered if and only if its projection to the first coordinate is partially well-ordered and its intersection with each fiber is partially well-ordered.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/WellFoundedSet.lean [hunk 1]
```diff
@@ -814,6 +815,97 @@ theorem partiallyWellOrderedOn_sublistForall₂ (r : α → α → Prop) [IsRefl
...
+        Exists.choose_spec (Exists.choose_spec hx)
+      aesop
+    · intro hx
+      simp_all only [mem_setOf_eq, mem_image, mem_inter_iff, mem_preimage, mem_singleton_iff]
+      exact Exists.intro (a, x) { left := { left := hx, right := rfl }, right := rfl }
+  rw [← h']
+  exact IsPWO.image_of_monotoneOn (IsPWO.mono hαβ <| inter_subset_left s ((fun x ↦ x.1) ⁻¹' {a})) h
+
+theorem ProdLex_iff [PartialOrder α] [Preorder β] {s : Set (α ×ₗ β)} :
+    s.IsPWO ↔ ((fun (x : α ×ₗ β) => x.1)'' s).IsPWO ∧
```
Comment (eric-wieser) [inline Mathlib/Order/WellFoundedSet.lean:900]:
```suggestion
    s.IsPWO ↔ ((fun (x : α ×ₗ β) => (ofLex x).1) '' s).IsPWO ∧
```
is better-typed

File: Mathlib/Order/WellFoundedSet.lean [hunk 2]
```diff
@@ -814,6 +815,97 @@ theorem partiallyWellOrderedOn_sublistForall₂ (r : α → α → Prop) [IsRefl
...
+      constructor
+      · exact (hhc (g' 0)).symm.trans (hhc (g' 1))
+      · exact h zero_le_one
+
+theorem imageProdLex [PartialOrder α] [Preorder β] {s : Set (α ×ₗ β)}
+    (hαβ : s.IsPWO) : ((fun (x : α ×ₗ β) => x.1)'' s).IsPWO :=
+  IsPWO.image_of_monotone hαβ Prod.Lex.monotone_fst
+
+theorem fiberProdLex [PartialOrder α] [Preorder β] {s : Set (α ×ₗ β)}
+    (hαβ : s.IsPWO) (a : α) : {y | (a, y) ∈ s}.IsPWO := by
```
Comment (eric-wieser) [inline Mathlib/Order/WellFoundedSet.lean:866]:
```suggestion
    (hαβ : s.IsPWO) (a : α) : {y | toLex (a, y) ∈ s}.IsPWO := by
```

Comment (eric-wieser) [reply Mathlib/Order/WellFoundedSet.lean:866]: This probably applies in quite a few places; when using `(_, _)` you need `toLex`, and when using `.1` or `.2` on `α ×ₗ β` you should have `ofLex`.

File: Mathlib/Data/Prod/Lex.lean [hunk 3]
```diff
@@ -104,6 +104,12 @@ instance preorder (α β : Type*) [Preorder α] [Preorder β] : Preorder (α ×
               exact h }
 #align prod.lex.preorder Prod.Lex.preorder
 
+theorem monotone_fst [Preorder α] [LE β] (t c: (α ×ₗ β)) (h : t ≤ c) :
+    (ofLex t).1 ≤ (ofLex c).1 := by
+  cases ((Prod.Lex.le_iff t c).mp h) with
+  | inl h' => exact LT.lt.le h'
+  | inr h' => exact le_of_eq h'.1
```
Comment (eric-wieser) [inline Mathlib/Data/Prod/Lex.lean:111]:
```suggestion
  | inl h' => exact h'.le
  | inr h' => exact h'.1.le
```

File: Mathlib/Data/Prod/Lex.lean [hunk 4]
```diff
@@ -104,6 +104,12 @@ instance preorder (α β : Type*) [Preorder α] [Preorder β] : Preorder (α ×
               exact h }
 #align prod.lex.preorder Prod.Lex.preorder
 
+theorem monotone_fst [Preorder α] [LE β] (t c: (α ×ₗ β)) (h : t ≤ c) :
```
Comment (eric-wieser) [inline Mathlib/Data/Prod/Lex.lean:107]: nit:
```suggestion
theorem monotone_fst [Preorder α] [LE β] (t c : α ×ₗ β) (h : t ≤ c) :
```

File: Mathlib/RingTheory/HahnSeries/Basic.lean [hunk 5]
```diff
@@ -120,41 +120,60 @@ theorem support_eq_empty_iff {x : HahnSeries Γ R} : x.support = ∅ ↔ x = 0 :
...
-    ∃ g : ℕ ↪o ℕ, Monotone (f ∘ g) :=
-  h.exists_monotone_subseq f hf
-#align set.is_pwo.exists_monotone_subseq Set.IsPWO.exists_monotone_subseq
-map sequence to Γ, get monotone subsequence (use ext property)
-if stationary at a ∈ Γ, look inside {a} × Γ', get monotone subsequence.
-if not stationary, use lex order.
+      exact (x.coeff a).isPWO_support'
+
 /-- Change a Hahn series on a lex product to a Hahn series with coefficients in a Hahn series. -/
 def to_iterate {Γ' : Type*} [PartialOrder Γ'] (x : HahnSeries (Γ ×ₗ Γ') R) :
```
Comment (eric-wieser) [inline Mathlib/RingTheory/HahnSeries/Basic.lean:142]:
```suggestion
def toIterate {Γ' : Type*} [PartialOrder Γ'] (x : HahnSeries (Γ ×ₗ Γ') R) :
```

File: Mathlib/RingTheory/HahnSeries/Basic.lean [hunk 6]
```diff
@@ -120,41 +120,60 @@ theorem support_eq_empty_iff {x : HahnSeries Γ R} : x.support = ∅ ↔ x = 0 :
...
+        exact rfl
+    rw [h₁]
+    have h : (Function.support fun g => fun g' => x.coeff (g, g')) =
+        ((fun x ↦ x.1) '' Function.support x.coeff) := by
+      exact Function.support_on_image (fun g => x.coeff g)
+    rw [h]
+    exact Set.PartiallyWellOrderedOn.imageProdLex x.isPWO_support'
+
+/-- The equivalence between iterated Hahn series and Hahn series on the lex product. -/
+def iterate_equiv {Γ' : Type*} [PartialOrder Γ'] :
```
Comment (eric-wieser) [inline Mathlib/RingTheory/HahnSeries/Basic.lean:171]:
```suggestion
@[simps]
def iterateEquiv {Γ' : Type*} [PartialOrder Γ'] :
```
which will automatically unfold it to `ofIterate` and `toIterate`

File: Mathlib/RingTheory/HahnSeries/Basic.lean [hunk 7]
```diff
@@ -120,41 +120,60 @@ theorem support_eq_empty_iff {x : HahnSeries Γ R} : x.support = ∅ ↔ x = 0 :
   support_eq_empty_iff.trans coeff_fun_eq_zero_iff
 #align hahn_series.support_eq_empty_iff HahnSeries.support_eq_empty_iff
 
-/-!
 /-- Change a HahnSeries with coefficients in HahnSeries to a HahnSeries on the Lex product. -/
 def of_iterate {Γ' : Type*} [PartialOrder Γ'] (x : HahnSeries Γ (HahnSeries Γ' R)) :
```
Comment (eric-wieser) [inline Mathlib/RingTheory/HahnSeries/Basic.lean:124]:
```suggestion
def ofIterate {Γ' : Type*} [PartialOrder Γ'] (x : HahnSeries Γ (HahnSeries Γ' R)) :
```

**Review by eric-wieser — COMMENTED** [review]
It looks like the commented out code wasn't fully lean4-ified

File: Mathlib/RingTheory/HahnSeries/Basic.lean [hunk 8]
```diff
@@ -120,41 +120,61 @@ theorem support_eq_empty_iff {x : HahnSeries Γ R} : x.support = ∅ ↔ x = 0 :
...
+    refine Set.PartiallyWellOrderedOn.subsetProdLex ?_ ?_
+    · have h : ((fun (x : Γ ×ₗ Γ') ↦ x.1) '' Function.support fun g ↦ (x.coeff g.1).coeff g.2) ⊆
+          Function.support x.coeff :=
+        Set.image_subset_iff.mpr <| support_subset_iff.mpr fun g hg => Set.mem_preimage.mpr <|
+        Function.mem_support.mpr <| ne_zero_of_coeff_ne_zero hg
+      exact Set.IsPWO.mono x.isPWO_support' h
+    · intro a
+      have h : {y | (a, y) ∈ Function.support fun g ↦ (x.coeff g.1).coeff g.2} =
+          Function.support fun b => (x.coeff a).coeff b := by
+        exact rfl
```
Comment (jcommelin) [inline Mathlib/RingTheory/HahnSeries/Basic.lean:137]:
```suggestion
```
Also, the `simp_all only ... exact` can be changed into `simpa only ... using`.

File: Mathlib/RingTheory/HahnSeries/Basic.lean [hunk 9]
```diff
@@ -120,41 +120,61 @@ theorem support_eq_empty_iff {x : HahnSeries Γ R} : x.support = ∅ ↔ x = 0 :
...
+      constructor
+      · intro y hy
+        simp_all only [Function.mem_support, ne_eq]
+        refine Not.intro ?left.h
+        rw [@HahnSeries.ext_iff]
+        simp only [imp_false, ne_eq]
+        exact hy
+      · intro y hy
+        simp_all only [Function.mem_support, ne_eq, not_not]
+        exact rfl
```
Comment (jcommelin) [inline Mathlib/RingTheory/HahnSeries/Basic.lean:162]: If you add
```
@[simp]
lemma mk_eq_zero (f : Γ → R) (h) : HahnSeries.mk f h = 0 ↔ f = 0 := by
  rw [HahnSeries.ext_iff]
  rfl
```
then this proof can become
```suggestion
      simp only [Function.support, ne_eq, mk_eq_zero]
```

File: Mathlib/RingTheory/HahnSeries/Basic.lean [hunk 10]
```diff
@@ -120,41 +120,61 @@ theorem support_eq_empty_iff {x : HahnSeries Γ R} : x.support = ∅ ↔ x = 0 :
...
+    rw [h₁]
+    have h : (Function.support fun g => fun g' => x.coeff (g, g')) =
+        ((fun x ↦ x.1) '' Function.support x.coeff) := by
+      exact Function.support_on_image (fun g => x.coeff g)
+    rw [h]
```
Comment (jcommelin) [inline Mathlib/RingTheory/HahnSeries/Basic.lean:167]:
```suggestion
    rw [h₁, Function.support_on_image x.coeff]
```

File: Mathlib/Algebra/Function/Support.lean [hunk 11]
```diff
@@ -220,6 +220,19 @@ theorem mulSupport_along_fiber_subset (f : α × β → M) (a : α) :
...
+    (mulSupport fun a => (fun b => f (a, b))) = (mulSupport f).image Prod.fst := by
+  refine mulSupport_eq_iff.mpr ?_
+  constructor
+  · intro x hx
+    simp_all only [mem_image, mem_mulSupport, ne_eq, Prod.exists, exists_and_right, exists_eq_right]
+    exact mulSupport_nonempty_iff.mp hx
+  · intro x hx
+    simp_all only [mem_image, mem_mulSupport, ne_eq, Prod.exists, exists_and_right, exists_eq_right,
+      not_exists, not_not]
+    exact rfl
```
Comment (jcommelin) [inline Mathlib/Algebra/Function/Support.lean:234]:
```suggestion
  simp [mulSupport, funext_iff, Set.image]
```

File: Mathlib/Order/WellFoundedSet.lean [hunk 12]
```diff
@@ -830,6 +831,101 @@ theorem partiallyWellOrderedOn_sublistForall₂ (r : α → α → Prop) [IsRefl
...
+    · by_contra hx
+      rw [Nat.not_lt] at hx
+      have hhc : (ofLex f (g 0)).1 < (ofLex f (g n)).1 := hc.choose_spec
+      have hhg : g n = g 0 := by
+        simp_all only [OrderEmbedding.le_iff_le, nonpos_iff_eq_zero]
+        exact congrArg (⇑g) hx
+      simp_all [hhg]
+    · refine (Prod.Lex.le_iff (f (g 0)) _).mpr ?_
+      left
+      exact hc.choose_spec
```
Comment (jcommelin) [inline Mathlib/Order/WellFoundedSet.lean:856]:
```suggestion
  · obtain ⟨n, hn⟩ := hc
    use (g 0), (g n)
    constructor
    · by_contra hx
      simp_all
    · refine (Prod.Lex.le_iff (f (g 0)) _).mpr <| Or.inl hn
```

**Review by jcommelin — COMMENTED** [review]
More to follow later

File: Mathlib/Order/WellFoundedSet.lean [hunk 13]
```diff
@@ -830,6 +831,101 @@ theorem partiallyWellOrderedOn_sublistForall₂ (r : α → α → Prop) [IsRefl
...
+      let b' : β := Exists.choose (Exists.choose_spec hx)
+      have hab' : (toLex (a', b') ∈ s ∧ (a', b').1 = a) ∧ (a', b').2 = x :=
+        Exists.choose_spec (Exists.choose_spec hx)
+      aesop
+    · intro hx
+      simp_all only [mem_setOf_eq, mem_image, mem_inter_iff, mem_preimage, mem_singleton_iff]
+      exact Exists.intro (toLex (a, x)) { left := { left := hx, right := rfl }, right := rfl }
+  rw [← h']
+  exact IsPWO.image_of_monotoneOn
+    (IsPWO.mono hαβ <| inter_subset_left s ((fun x ↦ (ofLex x).1) ⁻¹' {a})) h
```
Comment (jcommelin) [inline Mathlib/Order/WellFoundedSet.lean:917]:
```suggestion
  have h : {y | toLex (a, y) ∈ s} = f '' (s ∩ (fun x ↦ (ofLex x).1) ⁻¹' {a}) := by
    ext x
    simp [f]
  rw [h]
  apply IsPWO.image_of_monotoneOn (hαβ.mono (inter_subset_left s _))
  rintro b ⟨-, hb⟩ c ⟨-, hc⟩ hbc
  simp only [mem_preimage, mem_singleton_iff] at hb hc
  have : (ofLex b).1 < (ofLex c).1 ∨ (ofLex b).1 = (ofLex c).1 ∧ f b ≤ f c :=
    (Prod.Lex.le_iff b c).mp hbc
  simp_all only [lt_self_iff_false, true_and, false_or]
```

File: Mathlib/Order/WellFoundedSet.lean [hunk 14]
```diff
@@ -830,6 +831,101 @@ theorem partiallyWellOrderedOn_sublistForall₂ (r : α → α → Prop) [IsRefl
...
+
+theorem ProdLex_iff [PartialOrder α] [Preorder β] {s : Set (α ×ₗ β)} :
+    s.IsPWO ↔ ((fun (x : α ×ₗ β) => (ofLex x).1)'' s).IsPWO ∧
+    ∀ a, {y | toLex (a, y) ∈ s}.IsPWO := by
+  constructor
+  · intro h
+    constructor
+    · exact imageProdLex h
+    · exact fun a => fiberProdLex h a
+  · exact fun h => subsetProdLex h.1 h.2
```
Comment (jcommelin) [inline Mathlib/Order/WellFoundedSet.lean:927]:
```suggestion
    s.IsPWO ↔
      ((fun (x : α ×ₗ β) ↦ (ofLex x).1) '' s).IsPWO ∧ ∀ a, {y | toLex (a, y) ∈ s}.IsPWO :=
  ⟨fun h ↦ ⟨imageProdLex h, fiberProdLex h⟩, fun h ↦ subsetProdLex h.1 h.2⟩
```

File: Mathlib/Order/WellFoundedSet.lean (hunk 13 above)
Comment (jcommelin) [reply Mathlib/Order/WellFoundedSet.lean:906]: It looks like this suggestion still applies.

File: Mathlib/RingTheory/HahnSeries/Basic.lean [hunk 15]
```diff
@@ -120,41 +120,45 @@ theorem support_eq_empty_iff {x : HahnSeries Γ R} : x.support = ∅ ↔ x = 0 :
...
-  isPWO_support' := sorry
-  * Equivalence between `HahnSeries Γ (HahnSeries Γ' R)` and `HahnSeries (Γ × Γ') R`
-  * Use Set.WellFoundedOn.prod_lex_of_wellFoundedOn_fiber to iterate. (need PWO version)
--/
+  isPWO_support' := by
+    have h₁ : (Function.support fun g => HahnSeries.mk (fun g' => x.coeff (g, g'))
+        (Set.PartiallyWellOrderedOn.fiberProdLex x.isPWO_support' g)) = Function.support
+        fun g => fun g' => x.coeff (g, g') := by
+      simp only [Function.support, ne_eq, mk_eq_zero]
+    rw [h₁, Function.support_on_image x.coeff]
```
Comment (jcommelin) [inline Mathlib/RingTheory/HahnSeries/Basic.lean:151]:
```suggestion
    rw [h₁, Function.support_curry' x.coeff]
```

File: Mathlib/Algebra/Function/Support.lean [hunk 16]
```diff
@@ -244,6 +244,11 @@ theorem mulSupport_along_fiber_subset (f : α × β → M) (a : α) :
 #align function.mul_support_along_fiber_subset Function.mulSupport_along_fiber_subset
 #align function.support_along_fiber_subset Function.support_along_fiber_subset
 
+@[to_additive]
+theorem mulSupport_on_image (f : α × β → M) :
+    (mulSupport fun a => (fun b => f (a, b))) = (mulSupport f).image Prod.fst := by
+  simp [mulSupport, funext_iff, Set.image]
```
Comment (jcommelin) [inline Mathlib/Algebra/Function/Support.lean:250]:
```suggestion
@[to_additive]
theorem mulSupport_curry (f : α × β → M) :
    (mulSupport f.curry) = (mulSupport f).image Prod.fst := by
  simp [mulSupport, funext_iff, image]

@[to_additive]
theorem mulSupport_curry' (f : α × β → M) :
    (mulSupport fun a b ↦ f (a, b)) = (mulSupport f).image Prod.fst :=
  mulSupport_curry f
```

**Review by jcommelin — COMMENTED** [review]
LGTM

Thanks! :tada: 

bors d+

**Pull request closed without merging**